core/src/main/scala/cats/Traverse.scala

```diff
@@ -168,6 +168,48 @@ trait Traverse[F[_]] extends Functor[F] with Foldable[F] with UnorderedTraverse[
   def zipWithIndex[A](fa: F[A]): F[(A, Int)] =
     mapWithIndex(fa)((a, i) => (a, i))
 
+  /**
+    * Same as [[traverseWithIndexM]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def traverseWithLongIndexM[G[_], A, B](fa: F[A])(f: (A, Long) => G[B])(implicit G: Monad[G]): G[F[B]] =
+    traverse(fa)(a => StateT((s: Long) => G.map(f(a, s))(b => (s + 1, b)))).runA(0L)
+
+  /**
+    * Same as [[mapWithIndex]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def mapWithLongIndex[A, B](fa: F[A])(f: (A, Long) => B): F[B] =
+    traverseWithLongIndexM[cats.Id, A, B](fa)((a, long) => f(a, long))
+
+  /**
+    * Same as [[zipWithIndex]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def zipWithLongIndex[A](fa: F[A]): F[(A, Long)] =
+    mapWithLongIndex(fa)((a, long) => (a, long))
+
+  /**
+   * If `fa` contains the element at index `idx`, 
+   * return the version of `fa` where the element at `idx` is replaced with `b`. 
+   * If there is no element with such an index, return `None`. 
+   *
+   * The behavior is consistent with the Scala collection library's
+   * `updated` for collections such as `List`.
+   */
+  def updated[A, B >: A](fa: F[A], idx: Long, b: B): Option[F[B]] =
+    traverseWithLongIndexM[State[Boolean, *], A, B](fa) { case (elem, i) =>
```

Comment:
Isn't this going to be two levels deep of State nesting?

Can't we do one later deep by having a State that keeps the current index?

Reply:
Yeah, I think it is possible to do just a single traverse where the state keeps track of the index and if the element was updated or not, similar to how the `traverseWithLongIndex` was implemented.